### ISSUE TITLE
Feature request: ActivityPub support

### ISSUE THREAD

**Issue by lkdm** [Keats/gutenberg#2665]
# Read-only ActivityPub feeds for Zola

I created this issue to discuss the feasibility of adding a feature to Zola that serialises posts to `ActivityStream 2.0` format for ingestion by other federated instances.

## What is ActivityPub?

ActivityPub is a standard that allows different web applications to interact so information can be pooled across different instances.

It supports social networking, such as Mastodon or Lemmy, but can also be used for blogging in the case of [Plume](https://joinplu.me), [Write Freely](https://writefreely.org), and [Wordpress through a plugin](https://wordpress.org/plugins/activitypub/).

## Why

Why not use Plume or Write Freely? Because to my knowledge they are not statically rendered, and thus cost a lot more than running Zola locally and hosting the HTML on a free/cheap host.

Why ActivityPub? Because the new *Federated* internet runs on ActivityPub. Adding read-only `ActivityStream 2.0` support to Zola would allow social networks, or even other applications to subscribe to Zola blogs.

One of the *big* issues with self-hosting a blog is discoverability. A self-hosted blog from Zola is an island, unlike a blog post on Medium or Substack. Integrating with the *Fediverse* would help with this.

## How ActivityPub works

[ActivityPub works by establishing an `INBOX` and `OUTBOX` for an *actor*](https://en.wikipedia.org/wiki/ActivityPub#/media/File:ActivityPub-tutorial-image.svg).

Because Zola renders a static website, it is probably not feasible to implement an ActivityPub "Inbox" and therefore, comments.

However, an outbox activity stream could be published alongside the static HTML– so that federated instances may subscribe to Zola blogs.

## Format Examples

Here are some examples of `Activity Stream 2.0` objects that I found in the W3 specification that could be suitable:

### Post Detail endpoint

```json
{
      "@context": "https://www.w3.org/ns/activitystreams",
      "summary": "Martin added an article to his blog",
      "type": "Add",
      "published": "2015-02-10T15:04:55Z",
      "actor": {
       "type": "Person",
       "id": "http://www.test.example/martin",
       "name": "Martin Smith",
       "url": "http://example.org/martin",
       "image": {
         "type": "Link",
         "href": "http://example.org/martin/image.jpg",
         "mediaType": "image/jpeg"
       }
      },
      "object" : {
       "id": "http://www.test.example/blog/abc123/xyz",
       "type": "Article",
       "url": "http://example.org/blog/2011/02/entry",
       "name": "Why I love Activity Streams"
      },
      "target" : {
       "id": "http://example.org/blog/",
       "type": "OrderedCollection",
       "name": "Martin's Blog"
      }
    }
```

### Post List endpoint

```json
{
      "@context": "https://www.w3.org/ns/activitystreams",
      "summary": "Sally's posts",
      "type": "OrderedCollection",
      "totalItems": 2,
      "orderedItems": [
        {
          "type": "Post",
          "name": "Meeting 2016-11-17"
        },
        {
          "type": "Post",
          "name": "Reminder for Going-Away Party"
        }
      ]
    }
```

## Comment

I am somewhat familiar with Rust, so am happy to at some point commit some time to working on this (not until at least next year)– but I wanted to raise it first to check if it's feasible, desired, and worthwhile.

## Action items

- [ ] [ActivityPub plugin for WordPress](https://wordpress.org/plugins/activitypub/#description) apparently works well with Mastodon– it might be worth investigating the format the plugin publishes.

## Links

- [Introduction to ActivityPub](https://socialhub.activitypub.rocks/t/introduction-to-activitypub/508)
- [What is ActivityPub? – Wikipedia article](https://en.wikipedia.org/wiki/ActivityPub)
- [Activity Streams 2.0](https://www.w3.org/TR/activitystreams-core/)
- [Activity Streams Vocabulary](https://www.w3.org/TR/activitystreams-vocabulary/)

**Comment by Keats** [comment]
If I remember correctly someone wanted to do that before but it was not super practical. Maybe it's in the issues or the forum?

**Comment by lkdm** [comment]
@Keats I did have a *quick* search through the issues before posting and couldn’t find anything

**Comment by Keats** [comment]
Argh I swear it was talked about _somewhere_ but I can't remember

**Comment by kanru** [comment]
Someone did some investigation this year https://web.archive.org/web/20240317071325/https://maho.dev/2024/02/a-guide-to-implement-activitypub-in-a-static-site-or-any-website/

**Comment by kanru** [comment]
Another guide https://paul.kinlan.me/adding-activity-pub-to-your-static-site/

**Comment by lkdm** [comment]
That's a fantastic start, thank you

**Comment by Thatoo** [comment]
That's an awesome idea.
Did you have time to work on it @lkdm ? You were talking about 2025.

**Comment by lkdm** [comment]
@Thatoo Thanks for following up! Not at this stage. Life has been hectic for me due to wedding + honeymoon. I'm not currently doing any extracurricular coding. I'll keep this here though in case I get a chance.

**Comment by Popolon** [comment]
There is another post from 2020, even allowing comments cited (and improved) in this january 2024 (zola) blog post: https://fbanning.de/thoughts/mastodon-comments-on-zola-blog/

The original post in 2020: https://carlschwan.eu/2020/12/29/adding-comments-to-your-static-blog-with-mastodon/

**Comment by kanru** [comment]
I started this project with the goal to support any static site generator. https://github.com/pinka-social/pinka

It will be an active component, but it is kind of required to push updates to your followers.

**Comment by kanru** [comment]
I started this project with the goal to support any static site generator. https://github.com/pinka-social/pinka

It will be an active component, but it is kind of required to push updates to your followers.

**Comment by Thatoo** [comment]
That's a very cool project @kanru

**Comment by UnoYakshi** [comment]
[Duckquill](https://duckquill.daudix.one/) theme supports Mastodon comments. It also refers https://carlschwan.eu/2020/12/29/adding-comments-to-your-static-blog-with-mastodon/.

**Comment by Thatoo** [comment]
> [Duckquill](https://duckquill.daudix.one/) theme supports Mastodon comments. It also refers https://carlschwan.eu/2020/12/29/adding-comments-to-your-static-blog-with-mastodon/.

It looks nice and comments form mastodon is a cool feature but I didn't understand if it could also automatically publish a mastodon toot when a new article is released.

**Comment by Poolitzer** [comment]
No it doesn't, this would be more appropriate in the generator as well (since as OP points out this federation is basically a generated stream)

**Comment by Thatoo** [comment]
Someone tried/succeded to integrate pinka with Zola?

**Comment by Poolitzer** [comment]
>  but I didn't understand if it could also automatically publish a mastodon toot when a new article is released.

No, you enter the mastodon host, user and id manually in the blog posts.

**Comment by kanru** [comment]
> Someone tried/succeded to integrate pinka with Zola?

I haven't setup my personal blog with it, but I did hook it up on the Pinka blog https://pinka.dev/blog/posts/v0_1_0_release/

Feel free to start new discussions on the Pinka repo if you have any question.
